### ISSUE TITLE
ConnectionFailedError('Connection Refused: broker unavailable.') - how to recover?

### ISSUE THREAD

**Issue by mtin** [Azure/azure-iot-sdk-python#504]
Hi,

using azure-iot-device 2.1.0 I sometimes get device disconnects from which the client does not recover.

Other devices (connected to the same IoT-hub) continue to stream data. There are quite some network outages.

How should this be handled?

```bash
[azure.iot.device.common.pipeline.pipeline_ops_base][ERROR] 2020-03-19 14:32:01,946: ReauthorizeConnectionOperation: completing with error ConnectionFailedError('Connection Refused: broker unavailable.')
[azure.iot.device.common.pipeline.pipeline_stages_base][ERROR] 2020-03-19 14:32:01,947: ConnectionLockStage(ReauthorizeConnectionOperation): op failed.  Unblocking queue with error: ConnectionFailedError('Connection Refused: broker unavailable.')
[azure.iot.device.common.pipeline.pipeline_stages_base][ERROR] 2020-03-19 14:32:01,948: ConnectionLockStage(ReauthorizeConnectionOperation): failing MQTTPublishOperation op because of error
[azure.iot.device.common.pipeline.pipeline_ops_base][ERROR] 2020-03-19 14:32:01,948: MQTTPublishOperation: completing with error ConnectionFailedError('Connection Refused: broker unavailable.')
[azure.iot.device.common.pipeline.pipeline_ops_base][ERROR] 2020-03-19 14:32:01,948: SendD2CMessageOperation: completing with error ConnectionFailedError('Connection Refused: broker unavailable.')
[azure.iot.device.common.pipeline.pipeline_ops_base][ERROR] 2020-03-19 14:32:01,949: UpdateSasTokenOperation: completing with error ConnectionFailedError('Connection Refused: broker unavailable.')
[azure.iot.device.common.async_adapter][ERROR] 2020-03-19 14:32:01,949: Callback completed with error ConnectionFailedError('Connection Refused: broker unavailable.')
[azure.iot.device.common.async_adapter][ERROR] 2020-03-19 14:32:01,950: ["azure.iot.device.common.transport_exceptions.ConnectionFailedError: ConnectionFailedError('Connection Refused: broker unavailable.')\n"]
[asyncio][ERROR] 2020-03-19 14:32:01,951: Task exception was never retrieved
future: <Task finished coro=<Stream.publishing_async() done, defined at /deploy/streaming/stream.py:31> exception=ConnectionFailedError('Could not connect to IoTHub')>
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/azure/iot/device/iothub/aio/async_clients.py", line 30, in handle_result
    return await callback.completion()
  File "/usr/local/lib/python3.7/dist-packages/azure/iot/device/common/async_adapter.py", line 94, in completion
    return await self.future
azure.iot.device.common.transport_exceptions.ConnectionFailedError: ConnectionFailedError('Connection Refused: broker unavailable.')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/deploy/streaming/stream.py", line 35, in publishing_async
    await self._iot_client.publish(current_element, self.properties)
  File "/deploy/utils/azure_iot_client.py", line 36, in publish
    await self._client.send_message(message)
  File "<string>", line 1, in send_message
  File "/usr/local/lib/python3.7/dist-packages/azure/iot/device/iothub/aio/async_clients.py", line 158, in send_message
    await handle_result(callback)
  File "/usr/local/lib/python3.7/dist-packages/azure/iot/device/iothub/aio/async_clients.py", line 34, in handle_result
    raise exceptions.ConnectionFailedError(message="Could not connect to IoTHub", cause=e)
azure.iot.device.exceptions.ConnectionFailedError: ConnectionFailedError('Could not connect to IoTHub') caused by ConnectionFailedError('Connection Refused: broker unavailable.')
[azure.iot.device.iothub.pipeline.pipeline_stages_iothub][ERROR] 2020-03-19 14:32:01,949: UseAuthProviderStage(UpdateSasTokenOperation): token update operation failed.  Error=ConnectionFailedError('Connection Refused: broker unavailable.')
[azure.iot.device.common.handle_exceptions][ERROR] 2020-03-19 14:32:01,960: Exception caught in background thread.  Unable to handle.
[azure.iot.device.common.handle_exceptions][ERROR] 2020-03-19 14:32:01,960: ["azure.iot.device.common.transport_exceptions.ConnectionFailedError: ConnectionFailedError('Connection Refused: broker unavailable.')\n"]
```

AB#7366709

**Comment by jebrando** [comment]
@mtin Can you give us a snippet of the sample that you're using to product this error.

**Comment by mtin** [comment]
sure!
```python
    async def publishing_async(self):
        while not rospy.is_shutdown():
            if self._buffer.has_elements():
                await self._iot_client.publish(self._buffer.pop(), self.properties)
            else:
                await asyncio.sleep(Stream.PUBLISH_SLEEP_S)
```

**Comment by BertKleewein** [comment]
May be fixed in 2.1.1 which has significant improvements in this area, but I'm going to review your logs further.  Please let me know if this is resolved.

**Comment by mtin** [comment]
I got a little further, but still having troubles.

I realized I was swallowing exceptions from the azure iot client - so the `while` loop in my `publishing_async` coroutine would get canceled because of the iot client throwing an exception but I wouldn handle that.

**Comment by tanieee28** [comment]
Hi,
did we had any solution for the above problem ??
I am facing the same issue with **azure-iot-device-sdk-python - 2.1.1**

**Requirement:** I have written a python script using above sdk to send 180 json records in one minute.

**Problem:** the script runs smooth for some hours and send around 180 records, but at some time between it loses the connection to iot hub and hence, the record count sent to iot hub decreases.

Can i get any help on this ??

Below, are the logs for the time, when the python script lost the connection.

```
Live Generation : INFO : Total active threads : 10 Messaging Threads : 0
Live Generation : INFO : No of Records : 180. Timestamp : ['2020-05-06T20:33:00+00:00']
Live Generation : INFO : Average Message Delivery Time : 0.03502751736158735
Live Generation : ReauthorizeConnectionOperation: completing with error OperationCancelled('Transport timeout on connection operation')
Live Generation : ConnectionLockStage(ReauthorizeConnectionOperation): op failed.  Unblocking queue with error: OperationCancelled('Transport timeout on connection operation')
Live Generation : ConnectionLockStage(ReauthorizeConnectionOperation): failing MQTTPublishOperation op because of error
Live Generation : MQTTPublishOperation: completing with error OperationCancelled('Transport timeout on connection operation')
Live Generation : ConnectionLockStage(ReauthorizeConnectionOperation): failing MQTTPublishOperation op because of error
Live Generation : MQTTPublishOperation: completing with error OperationCancelled('Transport timeout on connection operation')
Live Generation : ConnectionLockStage(ReauthorizeConnectionOperation): failing MQTTPublishOperation op because of error
Live Generation : MQTTPublishOperation: completing with error OperationCancelled('Transport timeout on connection operation')
Live Generation : UpdateSasTokenOperation: completing with error OperationCancelled('Transport timeout on connection operation')
Live Generation : UseAuthProviderStage(UpdateSasTokenOperation): token update operation failed.  Error=OperationCancelled('Transport timeout on connection operation')
Live Generation : Exception caught in background thread.  Unable to handle.
Live Generation : ["azure.iot.device.common.pipeline.pipeline_exceptions.OperationCancelled: OperationCancelled('Transport timeout on connection operation')\n"]
Live Generation : INFO : Total active threads : 7 Messaging Threads : 0
Live Generation : ERROR : No of Records : 120. Timestamp : ['2020-05-06T20:34:00+00:00']
Live Generation : INFO : Average Message Delivery Time : 0.5325058757281694
```

**Comment by elhorton** [comment]
Hi @tanieee28 , apologies for the delay here. Are you still seeing this issue on version 2.1.4? We made several connection resiliency improvements in the past couple releases, could you please verify whether or not this is still an issue that requires investigation? Thank you for helping to improve our Python SDK.

**Comment by kunalg06** [comment]
can someone plz help me, getting smiliar errors. 
![image_2020_08_10T08_29_40_465Z](https://user-images.githubusercontent.com/18356081/89769665-233cab80-db1b-11ea-88f4-8fbda2a5f8be.png)
![image_2020_08_10T09_12_49_774Z](https://user-images.githubusercontent.com/18356081/89769679-30599a80-db1b-11ea-8888-3b60b0fde486.png)

**Comment by BertKleewein** [comment]
@kunalg06 - that error is not actually an error.  It is being logged as one, but the library is recovering and sending the message.  We will be changing the severity of that log message in a future release.

**Comment by kunalg06** [comment]
@BertKleewein Thank you for the reply, Due to this warning my code was getting stop after some time, maybe the library didnt got recovered. Can pls acknowledge me about this new issue i'm facing. Thanks in-advance. file:///home/kunal/Pictures/Screenshot%20from%202020-08-05%2012-03-22.png

**Comment by BertKleewein** [comment]
@kunalg06 - Can you please help me understand your situation more?  What version of azure-iot-device are you running, and what OS are you running it on?

I'm not sure what the logging you sent in your last message shows.  The azure-iot-device library does not do anything with JPEG decoding.  What is it about this log that shows that the library is not working?  Also, you mentioned a try/catch block, but I don't see anything in that log that indicates that you caught an error from our library.  What error is being caught?

**Comment by kunalg06** [comment]
@BertKleewein sorry for letting you know half baked things. 
1. corrupt jpeg errors are from opencv lib which I'm using in the script. It has nothing to do with azure-iot-hub lib. 
2. I'm using lib azure-iot-device version 2.1.3.
3. im sending an json object to azure iot hub, in some intervals. While sending after some time i receives these below msgs. And after some time script gets crashed down, so by using try-catch, I again call the method from IoTHubDeviceClient.create_from_connection_string() and initate the connection.
![89778947-0361b380-db2c-11ea-9564-7f18be627bfd](https://user-images.githubusercontent.com/18356081/91202511-8a637e00-e71f-11ea-8a00-21ae1043f445.png)
![89778958-08266780-db2c-11ea-9599-3c822672ebe5](https://user-images.githubusercontent.com/18356081/91202556-9b13f400-e71f-11ea-8874-9816c0a0aed3.png)
4. And sometimes i receive these belows errors also,
![image_2020_08_04T09_13_45_822Z](https://user-images.githubusercontent.com/18356081/91203515-c21ef580-e720-11ea-9428-17bbf562f6fc.png)
I just wanted to know what these msg indicates and how i can avoid it

**Comment by BertKleewein** [comment]
Can you share details on the exception that you catch?

The warning messages that you've shared can be ignored.  Connection problems like this are somewhat expected and the iot library is able to recover from them without loss of data.

Using a second call to create_from_connection_string as a recovery mechanism may cause problems though because it could lead to two instances of the IoTHubDevicClient object "fighting each other" for the single network connection.  This is why I think it's important to focus on the cause of that exception and see if we can eliminate that.

**Comment by elhorton** [comment]
Closing this issue due to lack of response, but if this remains an issue please re-open. Thanks!

**Comment by az-iot-builder-01** [comment]
@mtin, @BertKleewein, @tanieee28, @kunalg06, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by askpatrickw** [comment]
If you end up here, check if your ID_SCOPE matches what is on your /admin/device-connection settings page (assuming IoT Central there)

see also #670
